pkg/reconciler/taskrun/validate_resources.go

```diff
@@ -122,6 +122,53 @@ func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params
 		return fmt.Errorf("param types don't match the user-specified type: %s", wrongTypeParamNames)
 	}
 
+	return ValidateObjectParameters(paramSpecs, params)
+}
+
+// ValidateObjectParameters checks if all required keys of object type params are provided in taskrun params.
+// To use this function externally, make sure `paramSpecs` and `params` have
+// been checked against general missing/extra params and param type alignment,
+// which been done in `validateParams` in this module.
+func ValidateObjectParameters(paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param) error {
+	neededObjectParamKeys := make(map[string][]string)
+	providedObjectParamKeys := make(map[string][]string)
+
+	// collect needed keys for object parameters
+	for _, neededParamSpec := range paramSpecs {
+		if neededParamSpec.Type == v1beta1.ParamTypeObject {
+			for propertyKey := range neededParamSpec.Properties {
+				neededObjectParamKeys[neededParamSpec.Name] = append(neededObjectParamKeys[neededParamSpec.Name], propertyKey)
+			}
+		}
+	}
+
+	// collect provided keys for object parameters
+	for _, providedParam := range params {
+		if providedParam.Value.Type == v1beta1.ParamTypeObject {
+			for propertyKey := range providedParam.Value.ObjectVal {
+				providedObjectParamKeys[providedParam.Name] = append(providedObjectParamKeys[providedParam.Name], propertyKey)
+			}
+		}
+	}
+
+	// now collect all parameter names that have some/all keys missed in taskrun
+	missings := []string{}
+	for p, keys := range providedObjectParamKeys {
+		if _, ok := neededObjectParamKeys[p]; !ok {
+			// Ignore any missing params - this happens when extra params were
+			// passed to the task that aren't being used.
+			continue
+		}
+		missedKeys := list.DiffLeft(neededObjectParamKeys[p], keys)
+		if len(missedKeys) != 0 {
+			missings = append(missings, p)
+		}
```

Comment:
I realized the old comment gets outdated, but it shouldn't, but I think refactoring this part should be happening in this PR, rather than in the PR that results need it, as it is pretty straightforward that there is module that's needed in results side that a structure needs to be validated against ArraryOrString. ValidateObjectAgainstSchema(map[string]PropertySpec, object ArraryOrString) should be the one you should be looking for.

This should be a fairly quick restructure of the code that doesn't have to wait until results need to use this part. Let me know if you need help with that.

Reply:
Thought we agreed to further extract this helper function in next pr once result side gets clear requirement on this. (see the original thread)
I am cool with having this extracted in this pr. But there is some issues with the function signature

Reply:
> ValidateObjectAgainstSchema(map[string]PropertySpec, object ArraryOrString) should be the one you should be looking for.

Thought we agreed to extract this helper function further in next pr once result side gets clear requirement on this (see the original thread). I am cool with having this extracted in this pr and just tried with the function signature you just suggested. 
`map[string]PropertySpec, object ArraryOrString` implies we will only find missing keys for ***one pair*** of needed vs provided. That means we have to do something like the following to validate for all params/results of object type in both current module and where results use this, which seems not that elegant.
```
for name, key set in required:
  find the name's corresponding key set in provided:
      call this helper to check missing keys
```

So I just make this function to accept two maps `needed` and `provided`. The map key will be the object name, and corresponding value will be a list of keys for that object.

Let me know if this what you actually need on the result side. @Yongxuanzhang @ywluogg Happy to change otherwise.

Reply:
I think this function should work, thanks @chuangw6! Left few questions regarding some details.

Reply:
Actually I would prefer another implementation:
func ValidateObjectParameters(paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param)
1. Store all 
2. 

```
func ValidateObjectParameters(paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param) error {
	neededKeys := make(map[string][]string)
	// collect needed keys for object parameters
	for _, spec := range paramSpecs {
		if spec.Type == v1beta1.ParamTypeObject {
			for key := range spec.Properties {
				neededKeys[spec.Name] = append(neededKeys[spec.Name], key)
			}
		}
	}

	// collect provided keys for object parameters
	for _, p := range params {
		if p.Value.Type == v1beta1.ParamTypeObject {
			for key := range p.Value.ObjectVal {
				if err:=ValidateObjectAgainstSchema(neededKeys[p.Name], p.Value.ObjectVal);err!=nil{
                                       return err
                                  }
			}
		}
	}

func ValidateObjectAgainstSchema(needKeys[]string, object map[string]string){
     // one loop to store object keys in sets/map
    // one loop on needKeys and if not in sets/map return error

}
```

Reply:
I think this is related to this [comment](https://github.com/tektoncd/pipeline/pull/4786#discussion_r866935643). see response there. let me know if it doesn't make sense.

Reply:
I think Yongxuan's design for ValidateObjectAgainstSchema() would make more sense, as it gives flexibility that we don't have to bundle all params together. But what I was thinking is like this:
```
func ValidateObjectParameters(paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param) error {
       // this is a map of (pointers to a map)
	neededKeys := make(map[string]map[]*PropertySpec)
	// collect needed keys for object parameters
	for _, spec := range paramSpecs {
		if spec.Type == v1beta1.ParamTypeObject {
                         neededProperties[spec.Name] = &spec.Properties
		}
	}

	// validate param's objectval
	for _, p := range params {
		if p.Value.Type == v1beta1.ParamTypeObject {
				if err:=ValidateObjectAgainstSchema(neededKeys[p.Name], p.Value.ObjectVal);err!=nil{
                                       return err
			}
		}
	}

func ValidateObjectAgainstSchema(p map[]*PropertySpec, a *ArrayOrString){
     // if p is nil, means a is defined, but p doesn have it
    // if p is not nil, blah...

}
```

Reply:
what do you mean by bundling all params together? See this [comment](https://github.com/tektoncd/pipeline/pull/4786#discussion_r866935643) which is related to the design for this function.

Reply:
@ywluogg I think @chuangw6's current implementation can help to capture all missing keys and also aligned with code base, so it should be ok! We should move forward see if we can merge this pr asap

Reply:
I don't see the reason for merging this ASAP as a good reason to stay on a complicated design of piece of code.

To achieve capturing all the keys, the design can be the following:
```
func ValidateObjectParameters(paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param) error {
       // this is a map of (pointers to a map)
	neededKeys := make(map[string]*map[string]PropertySpec)
	// collect needed keys for object parameters
	for _, spec := range paramSpecs {
		if spec.Type == v1beta1.ParamTypeObject {
                         neededProperties[spec.Name] = &spec.Properties
		}
	}

      missed :=[]

	// validate param's objectval
	for _, p := range params {
		if p.Value.Type == v1beta1.ParamTypeObject {
				if m := ValidateObjectAgainstSchema(neededKeys[p.Name], p.Value); len(missedkeys) !=0 {
                                       missed = append(missed, m)
			}
		}
	}

      if len(missed) != 0 {
       // throw error together
      }

func ValidateObjectAgainstSchema(p *map[string]PropertySpec, a *ArrayOrString) []string {
     // if p is nil, means a is defined, but p doesnt have it
    // if p is not nil, blah...
  // at the end, return all missed keys for this single object
}
```